File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys


def is_safe(board, row, col, n):
    for i in range(row):
        if board[i] == col or \
                board[i] - i == col - row or \
                board[i] + i == col + row:
            return False
    return True


def count_n_queens_util(board, row, n, count):
    if row == n:
        count[0] += 1
        return

    for col in range(n):
        if is_safe(board, row, col, n):
            board[row] = col
            count_n_queens_util(board, row + 1, n, count)


def count_n_queens(n):
    count = [0]
    board = [-1] * n
    count_n_queens_util(board, 0, n, count)
    return count[0]


print(count_n_queens(int(sys.stdin.readline())))

File: 백준/Gold/9663. N－Queen/N－Queen.java
// N-Queen 9663

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
  private static int count = 0;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(br.readLine());
    br.close();

    countNQueens(n);
    System.out.println(count);
  }

  private static boolean isSafe(int[] board, int row, int col) {
    for (int i = 0; i < row; i++) {
      if (board[i] == col ||
          board[i] - i == col - row ||
          board[i] + i == col + row) {
        return false;
      }
    }
    return true;
  }

  private static void placeQueens(int[] board, int row, int n) {
    if (row == n) { // 모든 행에 퀸 배치.
      count++;
      return;
    }

    for (int col = 0; col < n; col++) {
      if (isSafe(board, row, col)) { // 각열에 퀸을 배치해보고 isSafe로 확인.
        board[row] = col;
        placeQueens(board, row + 1, n);
      }
    }
  }

  private static void countNQueens(int n) {
    int[] board = new int[n];
    placeQueens(board, 0, n);
  }
}
